File: src/banking/pages/withdraw_page.py
"""Withdraw page object for XYZ Banking Project."""
from playwright.sync_api import Page

from src.banking.helpers.browser_helpers import apply_ci_buffer, wait_for_angular
from src.banking.locators.selectors import CustomerDashboard, WithdrawPageSelectors


class WithdrawPage:
    def __init__(self, page: Page):
        self.page = page

    def open(self) -> None:
        """Click withdraw btn, wait for form."""
        self.page.locator(CustomerDashboard.withdraw_btn).click()
        wait_for_angular(self.page)
        self.page.wait_for_selector('form[ng-submit="withdrawl()"]', state="visible", timeout=5000)

    def enter_amount(self, amount: int) -> None:
        """Wait for amount input to be ready, then fill. Ensures form is stable before input."""
        loc = self.page.locator(WithdrawPageSelectors.amount_input)
        loc.wait_for(state="visible", timeout=5000)
        loc.fill(str(amount))

    def submit(self) -> None:
        """Click submit, wait for Angular to process, apply CI buffer for balance update."""
        btn = self.page.locator(WithdrawPageSelectors.submit_btn)
        btn.wait_for(state="visible", timeout=5000)
        btn.click()
        wait_for_angular(self.page)
        apply_ci_buffer(0.25)  # CI/CD: balance update; runners process slower

    def get_message(self) -> str:
        return (self.page.locator(WithdrawPageSelectors.message).text_content() or "").strip()

    def expect_success(self) -> None:
        msg = self.get_message()
        assert "Transaction successful" in msg

    def withdraw_amount(self, amount: int) -> None:
        """open → enter_amount → submit (no assertion)."""
        self.open()
        self.enter_amount(amount)
        self.submit()

File: playwright-ui/pages/WithdrawPage.ts
/**
 * Withdraw page object for XYZ Banking Project.
 * Mirrors src/banking/pages/withdraw_page.py.
 */
import type { Page } from '@playwright/test';
import { applyCiBuffer, waitForAngular } from './helpers';
import { CustomerDashboard, WithdrawPageSelectors } from '../locators/selectors';

export class WithdrawPage {
  constructor(private readonly page: Page) {}

  async open(): Promise<void> {
    await this.page.locator(CustomerDashboard.withdraw_btn).click();
    await waitForAngular(this.page);
    await this.page.waitForSelector('form[ng-submit="withdrawl()"]', { state: 'visible', timeout: 5000 });
  }

  async enterAmount(amount: number): Promise<void> {
    const loc = this.page.locator(WithdrawPageSelectors.amount_input);
    await loc.waitFor({ state: 'visible', timeout: 5000 });
    await loc.fill(String(amount));
  }

  async submit(): Promise<void> {
    const btn = this.page.locator(WithdrawPageSelectors.submit_btn);
    await btn.waitFor({ state: 'visible', timeout: 5000 });
    await btn.click();
    await waitForAngular(this.page);
    await applyCiBuffer(0.25);
  }

  async getMessage(): Promise<string> {
    return (await this.page.locator(WithdrawPageSelectors.message).textContent())?.trim() ?? '';
  }

  async expectSuccess(): Promise<void> {
    const msg = await this.getMessage();
    if (!msg.includes('Transaction successful')) {
      throw new Error(`Expected withdrawal success message, got: ${msg}`);
    }
  }

  /** open → enterAmount → submit (no assertion) */
  async withdrawAmount(amount: number): Promise<void> {
    await this.open();
    await this.enterAmount(amount);
    await this.submit();
  }
}
